### PR TITLE
Silently update Orbit's manual instrumentation to collect group id.

### PR DESCRIPTION
**Pull request by florian-kuebler**
The group id will be used to associate different timers to each other.
Currently, the data will only be collected and shown in the tooltip.

Note, that this is silently upgrading the version of manual
instrumentation. The public documentation (in the header) is not
yet updated.

Test: Profile OrbitTest (compiled with v0 and v1).
Bug: http://b/194764373

### CODE REVIEW

**Review by pierricgimmig — COMMENTED** [review]
Hi Florian, thanks for looking into this. I have some comments, mainly for code duplication in Orbit.cpp. I have a new version of that file based on yours, please see https://github.com/pierricgimmig/orbit/blob/feature/simplify_orbit_cpp/src/Api/Orbit.cpp. Let's discuss on Monday perhaps!?

File: src/Api/Orbit.cpp [hunk 1]
```diff
@@ -124,17 +159,32 @@ void orbit_api_set_enabled(uint64_t address, uint64_t api_version, bool enabled)
...
+    // By the time we reach this, the "initialized" guard variable has been set to 1 and we know
+    // that all function pointers have been written to and published to other cores by the use of
+    // acquire/release fences. The "enabled" flag serves as a global toggle which is always used in
+    // conjunction with the "initialized" flag to determine if the Api is active. See
+    // orbit_api_active() in Orbit.h.
+    api_v0->enabled = enabled;
+  }
+
+  orbit_api_v1* api_v1 = absl::bit_cast<orbit_api_v1*>(address);
+  if (!api_v1->initialized) {
```
[the remark this answers is not in the document]
Comment (florian-kuebler) [reply src/Api/Orbit.cpp]: There was a return missing. It should be either v0 or v1. I refactored the code, also to reduce duplication.

File: src/Api/Orbit.cpp [hunk 2]
```diff
@@ -124,17 +159,32 @@ void orbit_api_set_enabled(uint64_t address, uint64_t api_version, bool enabled)
         api_version, kOrbitApiVersion);
   }
 
-  orbit_api_v0* api_v0 = absl::bit_cast<orbit_api_v0*>(address);
-  if (!api_v0->initialized) {
-    // Note: initialize any newer api version before v0, which sets "initialized" to 1.
-    orbit_api_initialize_v0(api_v0);
+  if (api_version == 0) {
+    orbit_api_v0* api_v0 = absl::bit_cast<orbit_api_v0*>(address);
```
[the remark this answers is not in the document]
Comment (florian-kuebler) [reply src/Api/Orbit.cpp]: But isn't that essentially the point: We call (in EnableInTracee) `orbit_api_set_enabled(address, version, enabled)`
where address is the result from the call to `orbit_get_function_table_address_vi`, and `i == version`. So when we call this function with version 1, we have a pointer to an `orbit_api_v1` struct, when called with version 0, we get a pointer to an `orbit_api_v0` function table. Isn't that the whole point?

When someone is using the old Api header, we want to make sure, that we provide the functions to use for manual instrumentation with that version. And in this case the start function not taking the group id argument, as it is also not used at the call side, by the target.

**Review by florian-kuebler — COMMENTED** [review]
Thanks for the comment. I addressed most of the concerns, however, as suggested, lets discuss about the enable function.

File: src/Api/Orbit.cpp [hunk 3]
```diff
@@ -30,29 +38,21 @@ void EnqueueApiEvent(Types... args) {
...
 
 void orbit_api_start(const char* name, orbit_api_color color) {
-  // `__builtin_return_address(0)` will return us the (possibly encoded) return address of the
-  // current function (level "0" refers to this frame, "1" would be the callers return address and
-  // so on).
-  // To decode the return address, we call `__builtin_extract_return_addr`.
-  auto return_address =
-      absl::bit_cast<uint64_t>(__builtin_extract_return_addr(__builtin_return_address(0)));
-  EnqueueApiEvent<orbit_api::ApiScopeStart>(name, color, 0LU /*group_id*/, return_address);
+  auto return_address = ORBIT_GET_CALLER_PC();
```
Comment (pierricgimmig) [inline src/Api/Orbit.cpp:48]: auto -> uint64_t

File: src/Api/Orbit.cpp [hunk 4]
```diff
@@ -30,29 +38,21 @@ void EnqueueApiEvent(Types... args) {
...
 void orbit_api_stop() { EnqueueApiEvent<orbit_api::ApiScopeStop>(); }
 
 void orbit_api_start_async(const char* name, uint64_t id, orbit_api_color color) {
-  // `__builtin_return_address(0)` will return us the (possibly encoded) return address of the
-  // current function (level "0" refers to this frame, "1" would be the callers return address and
-  // so on).
-  // To decode the return address, we call `__builtin_extract_return_addr`.
-  auto return_address =
-      absl::bit_cast<uint64_t>(__builtin_extract_return_addr(__builtin_return_address(0)));
+  auto return_address = ORBIT_GET_CALLER_PC();
```
Comment (pierricgimmig) [inline src/Api/Orbit.cpp:55]: auto -> uint64_t, or use ORBIT_GET_CALLER_PC() directly in the function call maybe?

File: src/Api/Orbit.cpp [hunk 5]
```diff
@@ -87,15 +87,33 @@ void orbit_api_async_string(const char* str, uint64_t id, orbit_api_color color)
...
-static void orbit_api_initialize_v0(orbit_api_v0* api) {
-  // The api function table is accessed by user code using this pattern:
-  //
-  // bool initialized = api.initialized;
-  // std::atomic_thread_fence(std::memory_order_acquire)
-  // if (initialized && api->enabled && api->orbit_api_function_name) api->orbit_api_function_name()
-  //
-  // We use acquire and release semantics to make sure that when api->initialized is set, all the
-  // function pointers have been assigned and are visible to other cores.
+template <typename orbit_api_current>
```
Comment (pierricgimmig) [inline src/Api/Orbit.cpp:90]: nit. our templated types usually end with "T". Also, since this is c++, we should probably use camel case, OrbitApiT?

**Review by pierricgimmig — APPROVED** [review]
Thanks for the changes, lgtm, please see minor comments.

File: src/Api/Orbit.cpp (hunk 3 above)
Comment (florian-kuebler) [reply src/Api/Orbit.cpp]: Done.

File: src/Api/Orbit.cpp (hunk 4 above)
Comment (florian-kuebler) [reply src/Api/Orbit.cpp]: Done.

File: src/Api/Orbit.cpp (hunk 5 above)
Comment (florian-kuebler) [reply src/Api/Orbit.cpp]: Done.

**Pull request merged**